src/test/java/org/folio/fqm/service/EntityTypeFlatteningServiceTest.java

```diff
@@ -1,1075 +1,799 @@
+            "columns": [],
+            "sources": [
+              {
+                "type": "entity-type",
+                "alias": "simple_1",
+                "targetId": "00000000-0000-0000-0000-000000000001"
+              },
+              {
+                "type": "entity-type",
+                "alias": "simple_2",
+                "targetId": "00000000-0000-0000-0000-000000000002",
+                "targetField": "fieldZ",
+                "sourceField": "simple_1.field1"
+              }
+            ]
+          }
+          """,
+          EntityType.class
+        );
+      COMPOSITE_1W_TO_2W =
+        OBJECT_MAPPER.readValue(
+          """
+          {
+            "id": "eeeeeeee-0000-0000-0000-000000000102",
+            "name": "composite_1_wrapper_to_2_wrapper",
+            "columns": [],
+            "sources": [
+              {
+                "type": "entity-type",
+                "alias": "composite_1",
+                "targetId": "cccccccc-0000-0000-0000-000000000001"
+              },
+              {
+                "type": "entity-type",
+                "alias": "composite_2",
+                "targetId": "cccccccc-0000-0000-0000-000000000002",
+                "targetField": "simple_2.fieldZ",
+                "sourceField": "composite_1.simple_1.field1"
+              }
+            ]
+          }
+          """,
+          EntityType.class
+        );
+      COMPOSITE_COMPOSITE =
+        OBJECT_MAPPER.readValue(
+          """
+          {
+            "id": "ffffffff-0000-0000-0000-000000000000",
+            "name": "composite_composite",
+            "columns": [],
+            "sources": [
+              {
+                "type": "entity-type",
+                "alias": "composite_wrapper_to_wrapper",
+                "targetId": "eeeeeeee-0000-0000-0000-000000000102"
+              },
+              {
+                "type": "entity-type",
+                "alias": "composite_1_to_2",
+                "targetId": "dddddddd-0000-0000-0000-000000000012",
+                "targetField": "simple_2.fieldZ",
+                "sourceField": "composite_wrapper_to_wrapper.composite_1.simple_1.field1"
+              }
+            ],
+            "requiredPermissions": ["extra_perm"]
+          }
+          """,
+          EntityType.class
+        );
+      // what is that composite_composite amalgamation?
+      // composite_composite
+      // ├--- composite_wrapper_to_wrapper
+      // |    ├--- composite_1 <-----------┐ <------------------------┐ via composite_1 to simple_2 (D)
+      // |    |    └--- simple_1 -> db (A) | via simple_2 to simple_1 | resolves. simple_1 (A) to simple_2 (D)
+      // |    └--- composite_2 ------------┘                          |
+      // |         └--- simple_2 -> db (B)                            |
+      // └--- composite_1_to_2 ---------------------------------------┘
+      //      ├--- simple_1 -> db (C) ---┐
+      //      └--- simple_2 -> db (D) <--┘
```

Comment:
manmade horrors beyond human comprehension...but it's certainly good as a test case